[NO-TICKET] Update documentation to clarify what test mode does

ivoanjo opened the pull request:
**What does this PR do?**

This PR updates our documentation to clarify what test mode does:

1. I've tweaked the wording to make it clear that this feature is used for testing distributed tracing behavior

2. I've explicitly mentioned that traces will still be sent; the previous version of the docs erroneously claimed a no-op transport would be applied, which is not true

3. I've left hints for folks that search for "test" while trying to disable dd-trace-rb in their test suites, that they should look at `tracing.enabled` instead

4. I've removed the bigger code examples for this feature, as it's a very niche thing that almost nobody will want to use, and having the code examples invites copy-paste of something that is probably not what the customer wants

**Motivation:**

The fact this doc is confusing came up in #4149 so it seemed relevant to fix our docs to avoid more confusion in the future.

**Change log entry**

(Not relevant to changelog)

**Additional Notes:**

N/A

**How to test the change?**

Docs-only change :)

Strech commented on docs/GettingStarted.md:
````diff
@@ -2637,15 +2636,12 @@ or `c.agent.port`.
-  c.tracing.test_mode.enabled = true
-end
-```
+If you're looking to disable tracing while your test suite is running, consider using the `tracing.enabled` setting
+to disable sending of traces.
````
A minor thing, but it would be cool to give an example on how-to, even if it's obvious

```suggestion
If you're looking to disable tracing while your test suite is running, consider using the `tracing.enabled` setting
to disable sending of traces.

```ruby
Datadog.configure do |c|
  c.tracing.test_mode.enabled = false
end
```
```

ivoanjo replied:
Wait, I'm somewhat confused with this suggestion -- did you mean to suggest an example of using `tracing.enable`?

I explicitly removed the example with `test_mode.enabled` and stated why (item 4) so I'm not sure I understand your suggestion :eyes:

Strech reviewed (approved):
LGTM

p-datadog reviewed (approved):
Is test mode an internal feature not meant for use by customers at all?

ivoanjo commented:
> Is test mode an internal feature not meant for use by customers at all?

That's a good question. I've done some digging and this was added by @delner back in https://github.com/DataDog/dd-trace-rb/pull/1504 . So it seems like a vestige of a feature that was added to make CI work.

Perhaps something that we should look to deprecate for 3.x?

anmarchenko commented:
@p-datadog @ivoanjo test mode is actually used by `datadog-ci` library [like that](https://github.com/DataDog/datadog-ci-rb/blob/e8a361dd0098a4f7b82199fbdeb3726ea2920219/lib/datadog/ci/configuration/components.rb#L93-L96):

```ruby
          settings.tracing.test_mode.enabled = true
          settings.tracing.test_mode.async = true
          settings.tracing.test_mode.trace_flush = settings.ci.trace_flush || CI::TestVisibility::Flush::Partial.new
```

This is not supposed to be used by customers, so at some point we better remove it from the documentation completely.

ivoanjo commented:
Any thoughts on removing it from the doc in this PR? E.g. it's not like omitting docs is a breaking change ;)

anmarchenko commented:
@ivoanjo I am not against it, probably worth asking @marcotc about that

hestonhoffman commented on docs/GettingStarted.md:
```diff
@@ -2637,15 +2636,12 @@ or `c.agent.port`.
-When test mode is enabled, the tracer uses a `Test` adapter for no-op transport that can optionally buffer requests in
-test suites or other non-production environments. It is configured by setting `c.tracing.test_mode.enabled` to true.
-This mode only works for tracing.
+The test mode can be used for testing distributed tracing behavior. Enabling this setting will mean every
+span will be sent (no sampling is applied) and the transport will send data synchronously.
+It is configured by setting `c.tracing.test_mode.enabled` to true.
```
```suggestion
Use the test mode for testing distributed tracing behavior. Enabling this setting causes every
span to be sent (no sampling is applied) and causes the transport to send data synchronously.
To configure test mode, set `c.tracing.test_mode.enabled` to `true`.
```

hestonhoffman reviewed (commented):
Some minor style suggestions

p-datadog commented:
I like putting the internal settings under an `internal` namespace (same as what I have done for DI based on someone else's suggestion). Given that the setting is already implemented I would suggest just adding a "for internal DataDog use only" to the docs (in this PR).

marcotc commented:
> Any thoughts on removing it from the doc in this PR? E.g. it's not like omitting docs is a breaking change ;)

I agree here. Removing it from the public docs, then adding a comment in code on why we removed it (and that we should not expose it in the next major release).

ivoanjo commented:
Thanks folks! Here's what I've done:

1. Completely removed any references to `test_mode` from `GettingStarted.md`
2. Added a new section to `GettingStarted.md` to tell customers quickly how to disable tracing, and also merged it with the link to the CI gem
3. Removed user docs from the `settings.rb` definition, leaving only a note saying it's for internal use and that we may change it in the future

anmarchenko commented on docs/GettingStarted.md:
````diff
@@ -167,6 +167,24 @@ If your Agent runs on a different host or container than your application, or yo
+end
+```
+
+Furthermore, if you're trying to get more visibility into your tests or you're struggling with slow or flaky test suites, consider looking at Datadog's [Test optimization](https://docs.datadoghq.com/tests/) via the [`datadog-ci`](https://github.com/DataDog/datadog-ci-rb?tab=readme-ov-file#datadog-test-optimization-for-ruby) gem.
````
thanks for mentioning the test optimization tool!

ivoanjo replied:
Seemed very relevant to mention is folks are searching for "spec"/"test" in our readme :)

ivoanjo commented:
Going to go ahead and merge this one in, thanks everyone for the great feedback -- I think it's one less sharp edge that we just sanded off ;)
